SignalServiceKit/src/Util/DataSource.m

```diff
@@ -0,0 +1,280 @@
+
+- (nullable NSURL *)dataUrl
+{
+    NSString *_Nullable path = [self dataPath];
+    return (path ? [NSURL fileURLWithPath:path] : nil);
+}
+
+- (nullable NSString *)dataPath
+{
+    OWSAssert(self.dataValue);
+
+    @synchronized(self)
+    {
+        if (!self.cachedFilePath) {
+            NSString *dirPath = NSTemporaryDirectory();
+            NSString *fileName = [[[NSUUID UUID] UUIDString] stringByAppendingPathExtension:self.fileExtension];
+            NSString *filePath = [dirPath stringByAppendingPathComponent:fileName];
+            DDLogError(@"%@ ---- writing data", self.tag);
+            if ([self writeToPath:filePath]) {
+                self.cachedFilePath = filePath;
+            } else {
+                OWSFail(@"%@ Could not write data to disk: %@", self.tag, self.fileExtension);
+            }
+        }
+
+        return self.cachedFilePath;
+    }
+}
+
+- (nullable NSString *)dataPathIfOnDisk
+{
+    return self.cachedFilePath;
+}
+
+- (NSUInteger)dataLength
+{
+    OWSAssert(self.dataValue);
+
+    return self.dataValue.length;
+}
+
+- (BOOL)writeToPath:(NSString *)dstFilePath
+{
+    OWSAssert(self.dataValue);
+
+    // There's an odd bug wherein instances of NSData/Data created in Swift
+    // code reliably crash on iOS 9 when calling [NSData writeToFile:...].
+    // We can avoid these crashes by simply copying the Data.
+    NSData *dataCopy = (SYSTEM_VERSION_GREATER_THAN_OR_EQUAL_TO(10, 0) ? self.dataValue : [self.dataValue copy]);
+
+    BOOL success = [dataCopy writeToFile:dstFilePath atomically:YES];
+    if (!success) {
+        OWSFail(@"%@ Could not write data to disk: %@", self.tag, dstFilePath);
+        return NO;
+    } else {
+        return YES;
+    }
+}
+
+#pragma mark - Logging
+
++ (NSString *)tag
+{
+    return [NSString stringWithFormat:@"[%@]", self.class];
+}
+
+- (NSString *)tag
+{
+    return self.class.tag;
+}
+
+@end
+
+#pragma mark -
+
+@interface DataSourcePath ()
+
+@property (nonatomic) NSString *filePath;
+
+// These properties are lazy-populated.
+@property (nonatomic) NSData *cachedData;
+@property (nonatomic) NSNumber *cachedDataLength;
+
+@property (nonatomic, nullable) NSString *sourceFilename;
+
+@end
+
+#pragma mark -
+
+@implementation DataSourcePath
+
++ (nullable id<DataSource>)dataSourceWithURL:(NSURL *)fileUrl;
+{
+    OWSAssert(fileUrl);
+
+    if (!fileUrl || ![fileUrl isFileURL]) {
+        return nil;
+    }
+    DataSourcePath *instance = [DataSourcePath new];
+    instance.filePath = fileUrl.path;
+    return instance;
+}
+
++ (nullable id<DataSource>)dataSourceWithFilePath:(NSString *)filePath;
+{
+    OWSAssert(filePath);
+
+    if (!filePath) {
+        return nil;
+    }
+
+    DataSourcePath *instance = [DataSourcePath new];
+    instance.filePath = filePath;
+    return instance;
+}
+
+- (NSData *)data
+{
+    OWSAssert(self.filePath);
+
+    @synchronized(self)
+    {
+        if (!self.cachedData) {
+            DDLogError(@"%@ ---- reading data", self.tag);
```

Comment:
Similarly, why is this an error log - it seems normal?

Reply:
Removed.